### ISSUE TITLE
Saga Compatibility

### ISSUE THREAD

**Issue by allpro** [ctrlplusb/easy-peasy#219]
**Q**: I plan to implement EasyPeasy ('EP') inside a major app using Saga plus all the standard Redux boilerplate. If I implement EP in this site, can I be **100% confident** that existing Redux+Saga code will work as before? Are there any scenarios I need to watch for? If so, what are they?

I recommend the EP docs _explicitly_ discussed working _in parallel_ with other middleware/wrappers. The docs 'mention' that EP exports the _real_ Redux store, and that middleware _should_ work, but this is just a short paragraph. I suggest adding a dedicated page for this topic, to give people like me the confidence required before I add EP to a critical enterprise app.

Thanks in advance.
/Kevin

**Comment by ctrlplusb** [comment]
Hey @allpro 

I haven't touched Redux Saga in ages, so I wouldn't be able to say honestly right now. In principal Easy Peasy does support Redux middleware but there will be restrictions IMO. At the end of the day the underlying reducer format may not be what you expect. I'd have to have a play around with Easy Peasy + Saga.

Perhaps someone else here has already tried?

**Comment by allpro** [comment]
Thanks Sean,

The app is large, complex, and had a lot of Saga+Redux boilerplate.I want to transition to a cleaner structure, including a Redux wrapper. 

There will be an extended period where old code must COEXIST PEACEFULLY with new code, including new Redux branches/slices. This is what I was asking about.

All old reducers, etc., will be moved into a store model, - rewritten and simplified.

I'm committed to trying a proof of concept. If minor issues exist, I can patch them, otherwise not sure. Will post the  results... soon.

**Comment by ctrlplusb** [comment]
Sure thing, shout if you have any questions/blockers.

**Comment by allpro** [comment]
**_So far_ I have integrated Easy-Peasy (EP) with ordinary Redux actions, thunks and reducers.** 

See my detailed comments in **Issue https://github.com/ctrlplusb/easy-peasy/issues/222**

The test app can be pulled from: **https://github.com/allpro/easy-peasy-tester**

**My _next test_ will be with Redux Saga middleware.** Fingers-crossed!

**Comment by allpro** [comment]
**I have been using Easy-Peasy with Saga _and various other middleware_ for over a week, using the same reducerEnhancer described previously. I can confirm that everything works fine, for anyone interested in this.**

I am fighting an issue with Webpack hot-reloading, due to a convoluted integration with old Ember code. However I have confirmed this is a different issue that has nothing to do with Saga or other middleware integration.

**Comment by ctrlplusb** [comment]
@allpro would you be open to creating a recipe within the recipe section of the website?

**Comment by allpro** [comment]
> would you be open to creating a recipe 

Sure. I may not get to it for a few days though.

**Comment by ctrlplusb** [comment]
no rush. thanks

**Comment by vishalkrv** [comment]
@allpro It will be appreciated, if we can get a recipe of your usage of Saga with Easy Peasy.

**Comment by allpro** [comment]
@vishalkrv There isn't really anything required for EP to work along-side Saga, other than to combine the reducers and include the Saga middleware when you run createStore. I though there might be some issues with the Saga middleware, but there isn't. 

EP and Saga don't interoperate - they work in parallel, each updating the same Redux state with their own reducers. There should be some recipes in the docs for integrating the reducers.

My issue was the need to maintain the old reducers as-is, but that is not Saga-specific. To solve this, I made a special reducer handler using an EP method that allows me to call multiple 'root reducers'. I think I posted that design in another comment thread, but not sure where? If this is your challenge, I could repost that here.

**Comment by tonilaukka** [comment]
@allpro we have a similar situation but our boilerplate is originally react-boilerplate which has been recently turned to use create-react-app for easier update path. We rely heavily on redux and redux-sagas. A bit more detailed guide how to gradually migrate to EP would be greatly appreciated.

**Comment by allpro** [comment]
Hi @tonilaukka 

> We rely heavily on redux and redux-sagas. A bit more detailed guide how to gradually migrate to EP would be greatly appreciated.

In another issue, Sean provides an example for replacing your current root-reducer, (that uses combineReducers), with keys in your Easy-Peasy store-models that use your existing reducers. This is the simplest option.

In my scenario, the old root reducer was huge and had added complexity. Therefore I chose a method that runs every action through _both_ the old and new root-reducers _sequentially_. This meant I could leave the old root-reducer as-is. This has worked fine.

An example of my method is in the same comment thread as Sean's example: Issue #222

Another trick I use is to create new EP models for state-data when writing _new code_ that needs to read or interact with that data. However, I leave the _actual_ data in its old location in state, because old code and reducers expect it there. The new code accessed the data via a _computed value_ in the new model, which points to the old state location. This let me use new models and methods for all new code, while retaining backwards compatibility with old code using the same state. Other computed values in the model can read the 'alias' value in the model as if were the real data. 

When _everything_ using the old state value is migrated, I can remove it and change my new state value from a computed/alias value to be the _real_ state value, by just updating the reducer or action that loads it.

This pattern has helped to transitioning state that is widely used. We have a number of old 'reusable Redux pattern helpers', which many complex components utilize. It is therefore difficult to change our state structure!

**Comment by JoostSlijkoort** [comment]
For anyone (like me) coming to this looking how to hook up EP to Saga, it's as easy as this:

```
import { createStore } from 'easy-peasy';
import createSagaMiddleware from 'redux-saga'
import rootSaga from './store/rootSaga';

const sagaMiddleware = createSagaMiddleware();

const store = createStore(
  model,
{
  middleware: [sagaMiddleware],
});

sagaMiddleware.run(rootSaga);

export default store;
```

**Comment by DiegoBM** [comment]
We are going to start a new project and we were considering using Sagas, due to it's popularity, adoption, and features, but I recently I was informed about Easy Peasy, which also seems to be getting a lot of traction, but I have only found people talking about it in the context of having it working with something else, and having EP doing minimal things. So I was wondering, from the perspective of a large code base, are there any features/functionalities that you'd miss from Saga, if you were to use only EP? Is there something that you could do with saga, that you could not do with easy peasy?

@allpro are you going to fully replace Saga with Easy Peasy?

**Comment by ctrlplusb** [comment]
I am closing this. I feel like it is searchable within the GitHub issues and there is not enough demand to warrant inclusion within the official website. 💜
